algorithms/linfa-preprocessing/src/count_vectorization.rs

```diff
@@ -0,0 +1,491 @@
+            properties: VectorizerProperties {
+                convert_to_lowercase: true,
+                split_regex: r"\b\w\w+\b".to_string(),
+                n_gram_range: (1, 1),
+                normalize: true,
+                document_frequency: (0., 1.),
+            },
+        }
+    }
+}
+
+impl CountVectorizer {
+    pub fn convert_to_lowercase(mut self, convert_to_lowercase: bool) -> Self {
+        self.properties.convert_to_lowercase = convert_to_lowercase;
+        self
+    }
+
+    pub fn split_regex(mut self, regex_str: &str) -> Self {
+        self.properties.split_regex = regex_str.to_string();
+        self
+    }
+
+    pub fn n_gram_range(mut self, min_n: usize, max_n: usize) -> Self {
+        self.properties.n_gram_range = (min_n, max_n);
+        self
+    }
+
+    pub fn normalize(mut self, normalize: bool) -> Self {
+        self.properties.normalize = normalize;
+        self
+    }
+
+    pub fn document_frequency(mut self, min_freq: f32, max_freq: f32) -> Self {
+        self.properties.document_frequency = (min_freq, max_freq);
+        self
+    }
+
+    /// Learns a vocabulary from the texts in `x`, according to the specified attributes and maps each
+    /// vocabulary entry to an integer value, producing a [FittedCountVectorizer](struct.FittedCountVectorizer.html).
+    ///
+    /// Returns an error if:
+    /// * one of the `n_gram` boundaries is set to zero or the minimum value is greater than the maximum value
+    /// * if the minimum document frequency is greater than one or than the maximum frequency, or if the maximum frequecy is  
+    ///   smaller than zero
+    /// * if the regex expression for the split is invalid
+    pub fn fit<T: ToString + Clone, D: Data<Elem = T>>(
+        &self,
+        x: &ArrayBase<D, Ix1>,
+    ) -> Result<FittedCountVectorizer> {
+        let regex = validate_properties(&self.properties)?;
+
+        // word, (integer mapping for word, document frequency for word)
+        let mut vocabulary: HashMap<String, (usize, usize)> = HashMap::new();
+        for string in x
+            .iter()
+            .map(|s| transform_string(s.to_string(), &self.properties))
+        {
+
+            let mut document_vocabulary: HashSet<String> = HashSet::new();
+            let words = regex.find_iter(&string).map(|mat| mat.as_str()).collect();
+            let queue = NGramQueue::new(words, self.properties.n_gram_range);
+            for ngram_items in queue {
+                for item in ngram_items {
+                    // if item was already in the hashet it simply gets overwritten,
+                    // not a problem
+                    document_vocabulary.insert(item);
+                }
+            }
+            for word in document_vocabulary {
+                let len = vocabulary.len();
+                // If vocabulary item was already present then increase its document frequency
+                if let Some((_, freq)) = vocabulary.get_mut(&word) {
```

Comment:
`document_vocabulary` is a hash-set but `vocabulary` is a hash-map. `document_vocabulary` is different for each document, so a vocabulary item may be already in `vocabulary` because it appeared in a previous document

Reply:
you're right, haven't seen the outer loop